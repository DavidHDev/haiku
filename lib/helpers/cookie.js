export const parseToDataType = (value, isItRetry = false) => {
  try {
    return value === 'undefined' ? undefined : JSON.parse(value ?? '');
  } catch (e) {
    if (!isItRetry) return parseToDataType(`"${value?.replaceAll?.('"', '')}"`, true);
    return undefined;
  }
};

export const parseToCookieType = (value) => {
  if(typeof value === 'string') return value;

  return JSON.stringify(value);
}

export const getCookie = (name) => {
  const value = `; ${document.cookie}`;

  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType(cookie.split(';')[0]) : cookie;
};

export const getCookies = (cookies = []) => {
  if (cookies.length) return cookies.reduce((result, cookie) => ({ ...result, [cookie]: getCookie(cookie) }), {});

  return Object.fromEntries(document.cookie.split('; ').map((c) => {
    const [key, value] = c.split('=');
    return [key, parseToDataType(value)]
  }));
};

export const setCookie = (name, value, expireDays) => {
  const date = new Date();

  date.setTime(date.getTime() + expireDays * 24 * 60 * 60 * 1000);
  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name) => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
