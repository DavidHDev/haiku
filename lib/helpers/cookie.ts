export const parseToDataType = <T>(
  value: string | undefined,
  isItRetry = false,
): T | undefined => {
  try {
    return value === 'undefined' || value === undefined
      ? undefined
      : (JSON.parse(value) as T);
  } catch (e) {
    if (!isItRetry) {
      return parseToDataType<T>(`"${value?.replaceAll?.('"', '')}"`, true);
    }
    return undefined;
  }
};

export const parseToCookieType = <T>(value: T): string => {
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
};

export const getCookie = <T>(name: string): T | undefined => {
  const value = `; ${document.cookie}`;
  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType<T>(cookie.split(';')[0]) : undefined;
};

export const getCookies = <T>(
  cookies: string[] = [],
): Record<string, T | undefined> => {
  if (cookies.length) {
    return cookies.reduce(
      (result, cookie) => ({ ...result, [cookie]: getCookie<T>(cookie) }),
      {} as Record<string, T | undefined>,
    );
  }

  return Object.fromEntries(
    document.cookie.split('; ').map((c) => {
      const [key, value] = c.split('=');

      return [key, parseToDataType<T>(value)];
    }),
  ) as Record<string, T | undefined>;
};

export const setCookie = <T>(
  name: string,
  value: T,
  expireDays: number,
): void => {
  const date = new Date();
  const millisecondsInADay = 24 * 60 * 60 * 1000;
  const expireDate = date.getTime() + expireDays * millisecondsInADay;

  date.setTime(expireDate);

  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name: string): void => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
