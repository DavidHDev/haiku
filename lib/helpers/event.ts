export const on = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.addEventListener)
    target.addEventListener(type, callback, options);
};

export const off = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.removeEventListener)
    target.removeEventListener(type, callback, options);
};
