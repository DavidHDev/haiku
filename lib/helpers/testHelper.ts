import { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { expect } from "vitest";

export const takeSnapshot = (ui: ReactNode) => {
  const { asFragment } = rtlRender(ui);
  expect(asFragment()).toMatchSnapshot();
};
