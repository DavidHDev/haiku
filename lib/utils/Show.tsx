import { Children, isValidElement, ReactNode } from 'react';
import { If } from './If';

type Props = {
  children: ReactNode;
};

type ElseProps = {
  render?: () => ReactNode;
  children?: ReactNode;
};

export const Show = ({ children }: Props) => {
  let when: ReactNode | null = null;
  let otherwise: ReactNode | null = null;

  Children.forEach(children, (child) => {
    if (isValidElement(child)) {
      if ('isTrue' in child.props) {
        if (!when && child.props.isTrue) when = child;
      } else otherwise = child;
    }
  });

  return when || otherwise;
};

Show.When = If;
Show.Else = ({ render, children }: ElseProps) => (render ? render() : children);
