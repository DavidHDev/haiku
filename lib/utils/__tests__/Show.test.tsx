import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { takeSnapshot } from 'lib/helpers/testHelper';
import { Show } from '../Show';

describe('Show', () => {
  const whenText = 'When Component';

  it('renders the Show component', () => {
    takeSnapshot(
      <Show>
        <Show.When isTrue>{whenText}</Show.When>
      </Show>,
    );
    expect(screen.getByText(whenText)).toBeInTheDocument();
  });

  it('renders the Else component when isTrue is false', () => {
    const elseText = 'Else Component';
    render(
      <Show>
        <Show.When isTrue={false}>{whenText}</Show.When>
        <Show.Else>{elseText}</Show.Else>
      </Show>,
    );
    expect(screen.queryByText(whenText)).not.toBeInTheDocument();
    expect(screen.getByText(elseText)).toBeInTheDocument();
  });

  it('ignores non-element children', () => {
    const inValidText = 'This is not a React element';
    const { queryByText } = render(
      <Show>
        {inValidText}
        <Show.When isTrue>{whenText}</Show.When>
      </Show>,
    );

    expect(queryByText(whenText)).toBeInTheDocument();
    expect(queryByText(inValidText)).toBeNull();
  });

  it('render render function when pass', () => {
    const renderText = 'Render text';
    const renderFuntion = () => renderText;
    const { queryByText } = render(
      <Show>
        <Show.When isTrue={false}>{whenText}</Show.When>
        <Show.Else render={renderFuntion} />
      </Show>,
    );

    expect(queryByText(renderText)).toBeInTheDocument();
  });
});
