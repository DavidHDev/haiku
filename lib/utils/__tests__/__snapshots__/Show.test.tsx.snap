// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Show > renders the Show component 1`] = `
<DocumentFragment>
  When Component
</DocumentFragment>
`;
